### DISCUSSION TITLE
Module-based development for a Large-Scale web application.

### DISCUSSION THREAD

**Discussion by EddyDezuraud** [nuxt/nuxt#25155] [Questions]
Hello everyone 👋,

I'm currently working on an large web application in Nuxt3 that allows managing a wide range of functionalities (form creation, configuration, real-time tracking, results, etc.).

**Reproduction sample : https://github.com/EddyDezuraud/nuxt-large-boilerplate/tree/master**

This application was initially developed in Nuxt2 and was recently migrated to Nuxt3. To clearly define each component, pages, etc ... of the application, we opted for the modular approach and created a module for each component.

For example, we have a dedicated module for real-time tracking of the number of visitors. Inside this module, there is a basic Nuxt structure, including pages, components, store, and so on.

Another aspect that doesn't seem standard in the Nuxt documentation is our decision to use services for API calls. We create a service folder for each module, which contains a list of calls using useFetch. This approach allows us to centralize all the calls in one place and have a clear list of data and calls that can be utilized for each module.

I've noticed some changes recently regarding the internal module aspect of the application, and it seems that it may no longer be considered a best practice in Nuxt. I've heard about "Layers" recently.

I would like to hear your opinion on the practice of adding internal modules, whether it's still considered a good practice or not. Also, I'm interested in any recommendations you might have on this topic.

**Comment by s0h311** [comment]
The concept of service is generally common in Angular and backend frameworks. I would create custom composables for things like API calls. For that you can create a `composables` directory. Composables handle reactive data and / or wrap other composables such as `useFetch`. If you are just extracting non-reactive functionality, just create helpers. Maybe you could orginze (or call) those "services". 

Consider investing a couple of minutes and watching a couple of vids from @manniL:

https://youtu.be/njsGVmcWviY?si=lTaHpP6NI6cxgJGu
https://youtu.be/N0QrFKBZuqA?si=BBC9uX6ws4WEOLYj
https://youtu.be/DkvgJa-X31k?si=OwQiwHCbviyr53d3

Please correct me, if I'm wrong.

**Reply by EddyDezuraud** [reply]
Thank you for sharing the videos and the areas for improvement.

I find the approach of using composables very interesting for this use case indeed. In our extensive application, we utilize a central composable, composables/fetch.ts, to streamline the handling of tokens and cookies during requests to the API.

The idea behind creating a /services folder was to centralize all API call URLs along with their types directly using a syntax like: "`ApiService.get('api/rest/admin/xxx')`". This makes communication and information sharing between the front and back-end teams working on the project more straightforward.

However, it might be worth reconsidering this approach and instead adopting the use of composables and describing API calls via a JSON description within each module. For example:
```
const API_CALLS: Record<string, { method: string, url: string, args?: object }> = {
  getUsers: { method: 'GET', url: '/api/users' },
  createUser: { method: 'POST', url: '/api/users', args: { name: '', email: '', password: '' } },
};
```

On your end, do you typically centralize all calls from an application module within a single file to facilitate communication between the front and back end? Or how do you deal with ?

**Reply by s0h311** [reply]
No matter how many Entities I have to fetch I always create separate composables for each. For example if I wanted to fetch a list of Products I would create a `useFetchProducts` composable that under the hood uses `useFetch`. I personally like separating the API calls, but if it makes sense logically and from a business case perspective you could also combine a few.

I must mention that I don't build any enterprise scale apps with Nuxt.

**Comment by TheAlexLichter** [comment]
I understand that splitting up the project into domains can help a lot. In addition to what @s0h311 suggested and you already mentioned, Nuxt Layers could help the most here, reassembling a typical nuxt application and separate logical concerns.

A base layer could hold general app utilities, and base components, "domain-specifiic" layers hold the content that is related to the concern and the "app" plugs all together.

![image](https://github.com/nuxt/nuxt/assets/640208/d2c89cde-7ad4-4982-8cdd-f560bd8ad39a)

**Reply by EddyDezuraud** [reply]
Thanks, Alexander, for your feedback. By the way, I absolutely love your Nuxt videos! 😍

I just took a look at the documentation regarding Layers. It could be extremely interesting for us to externalize certain large modules into separate projects and then reintegrate them into the central project.

However, we would still like the ability to keep some modules within the main project and simply have isolated module-specific folders like "components, pages, store, composables." 

As I explore this approach, it doesn't seem possible with the layers system, am I correct?

**Reply by TheAlexLichter** [reply]
@EddyDezuraud yes and no. You *can* do that with modules and "auto-import" utils, composables etc. etc. with a bit of extra work in the module. But you could also create a monorepo with these folders layers instead 😊

Thanks a lot - glad you enjoy my videos 🙏
